node/src/components/consensus/highway_core/state/endorsement_tracker.rs

```diff
@@ -0,0 +1,103 @@
+use std::collections::HashMap;
+
+use crate::components::consensus::{highway_core::validators::ValidatorIndex, traits::Context};
+
+use super::{Panorama, State};
+
+/// Helper structure used for computing which votes need endorsement.
+pub(crate) struct EndorsementTracker<'a, C: Context> {
+    /// Validators that are known to be equivocators by the initial panorama.
+    known_equivocators: Vec<ValidatorIndex>,
+    /// Visited units. Used for avoding visitng the same unit twice.
+    visited_nodes: Vec<C::Hash>,
+    /// Collection from a unit hash to indexes of equivocators that this unit cites as honest.
+    equiv_citations: HashMap<C::Hash, Vec<ValidatorIndex>>,
+    /// Read-only state.
+    state: &'a State<C>,
+}
+
+impl<'a, C: Context> EndorsementTracker<'a, C> {
+    pub(crate) fn new(state: &'a State<C>) -> Self {
+        EndorsementTracker {
+            known_equivocators: Vec::new(),
+            visited_nodes: Vec::new(),
+            equiv_citations: HashMap::new(),
+            state,
+        }
+    }
+
+    /// Returns votes that cite equivocating validators as honest.
+    pub(crate) fn run(mut self, panorama: &Panorama<C>) -> HashMap<C::Hash, Vec<ValidatorIndex>> {
+        self.known_equivocators = panorama.iter_faulty_validators().collect();
+        if self.known_equivocators.is_empty() {
+            return HashMap::new();
+        }
+        for v in panorama.iter_correct_hashes() {
+            self.find_equivocations(&v)
+        }
+        self.equiv_citations
+    }
+
+    /// Looks for equivocations in a downward set of vote.
+    fn find_equivocations(&mut self, vhash: &C::Hash) {
+        self.visited_nodes.push(*vhash);
+        let cited = self.cited_equivocators(vhash);
+        if !cited.is_empty() {
+            self.equiv_citations.insert(*vhash, cited);
+        }
+
+        // Units seen as correcy by panorama of `vhash`.
+        let to_visit: Vec<C::Hash> = self
+            .state
+            .vote(vhash)
+            .panorama
+            .iter_correct_hashes()
+            .filter(|&v| {
+                // Do not follow units that are created by known equivocators,
+                // have already been seen as endorsed by this node's state
+                // or has already been visited.
+                !self.by_equivocator(v) || !self.state.is_endorsed(v) || !self.already_visited(v)
```

Comment:
For _validity_, after all dependencies are satisfied, `state.is_endorsed` should be irrelevant. Whether a unit satisfies the LNC must not depend on our local state.

Reply:
It's not about unit satisfying the LNC but pruning branches of the DAG we don't have to visit.

Reply:
But those must not depend on our local latest state either.
(Sorry, I still don't understand this module well enough to have an opinion on what it _should_ say here.)

Reply:
Maybe what it _should_ check instead is whether the vote that `panorama` belongs to _claims_ that `v` is endorsed?

Reply:
Actually, I don't think we can stop the iteration at a unit by an equivocator either: That unit might still be part of a chain that constitutes an _indirect_ citation of otherwise unrelated units.

Reply:
> Actually, I don't think we can stop the iteration at a unit by an equivocator either: That unit might still be part of a chain that constitutes an _indirect_ citation of otherwise unrelated units.

I agree :grimacing:  I was thinking the same but wanted to first create a test.